### ISSUE TITLE
ASAN crashes in exiv2 on 'main' with the command `$ exiv2 -vVg exiv2` (on M1 Mac)

### ISSUE THREAD

**Issue by clanmills** [Exiv2/exiv2#2114]
I believe the crash occurs in the exiv2 application and not in the library.  Therefore I don't believe this is a security issue.  It was discovered and mentioned:  https://github.com/Exiv2/exiv2/pull/2110#issuecomment-1048831136

**Comment by mallman** [comment]
Note, I'm unable to reproduce this issue when I run exiv2 in x86_64 Rosetta mode on an M1 mac. I do see this issue with the arm64 build:

```
[msa@hoodoo exiv2]$ build/bin/exiv2 -pa test/data/_DSC8437.exv 
/Users/msa/coad/ImageMetadata/exiv2/app/actions.cpp:497:25: runtime error: member call on address 0x000109304f20 which does not point to an object of type 'Exiv2::Metadatum'
0x000109304f20: note: object is of type 'Exiv2::Exifdatum'
 01 00 00 00  90 40 fd 05 01 00 00 00  90 21 c0 04 01 00 00 00  c0 41 a0 04 01 00 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              vptr for 'Exiv2::Exifdatum'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/msa/coad/ImageMetadata/exiv2/app/actions.cpp:497:25 in 
Abort trap: 6
```

**Comment by clanmills** [comment]
This issue seems to be with Apple clang 13.0.0 on the arm64/m1/Apple Silicon platform.
https://github.com/Exiv2/exiv2/issues/2117#issuecomment-1050561782

@mallman  Can you reproduce my finding concerning the scope of this issue.

**Comment by clanmills** [comment]
The following test are on your branch with -DEXIV2_TEAM_USE_SANITIZERS=On

MacBookPro arm64/m1/AppleSilicon:

```bash
628 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ sw_vers
ProductName:	macOS
ProductVersion:	12.2
BuildVersion:	21D49
629 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ env DYLD_LIBRARY_PATH=$PWD/lib bin/exiv2 -vVg exiv2
exiv2 1.0.0.9
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/regex:5815:30: runtime error: member call on address 0x0001079006b0 which does not point to an object of type 'std::__node<char>'
0x0001079006b0: note: object is of type 'std::__1::__empty_state<char>'
 00 00 00 00  f8 14 4c 02 01 00 00 00  00 2b 80 07 01 00 00 00  02 02 00 00 08 00 00 00  03 00 00 7e
              ^~~~~~~~~~~~~~~~~~~~~~~
              vptr for 'std::__1::__empty_state<char>'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/regex:5815:30 in 
Abort trap: 6
630 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ 
```
MacMini x64/intel:

```bash
530 rmills@rmillsmm-local:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.7
BuildVersion:	19H1715
531 rmills@rmillsmm-local:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ env DYLD_LIBRARY_PATH=$PWD/lib bin/exiv2 -vVg exiv2
exiv2 1.0.0.9
exiv2=1.0.0
processpath=/Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build/bin
package_name=exiv2
executable=/Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build/bin/exiv2
library=/Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build/lib/libexiv2.1.0.0.9.dylib
config_path=/Users/rmills/.exiv2
532 rmills@rmillsmm-local:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ 
631 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $
```

I believe there's a "dirty hack" fix.  The crash is coming from regex.  I believe we can modify CMakeLists.txt to avoid using regex on arm64.  I'll investigate today (Friday) while you're sleeping.

**Comment by clanmills** [comment]
Because 'main' is C++17, it always uses `std::regex` for the `exiv2 -g` option.  The 0.27-maintenance CMake/Regex code to implement `-g` with regex.h or regex or sub-string-match has been _correctly_ removed. 

It is possible to totally avoid regex by changing src/version.cpp `shouldOuput()` to return true.  However as `exiv2 -g` is frequently used in the test suite, this evades the crash and causes massive collateral damage to the test suite.

The ASAN crash in #2115 is another `vptr` arm64 issue and has nothing to do with `shouldOutput()` or the `-g` option.

Both #2114 and #2115 seem to be restricted to branch 'main' on macOS/arm64.  I don't believe the CI builds this platform and that explains why these issues went undetected.  Until working with @mallman, I have seldom built `main`.

**Comment by clanmills** [comment]
@piponazo.  I looked at the regex documentation to see if I could do a sub-string match by obtaining the pattern from the regex.  I didn't find anything to help.

I didn't step the code in the debugger.  My suspicion is the Clang C++17 compiler is nervous about something involving the `vptr` as we also saw in #2115.  The code (for both this and 2115) seems to run correctly when ASAN is not enabled.

We could just sit on this to see if it disappears with a future release of Xcode bring a modified macOS/SDK.

**Comment by mallman** [comment]
> @clanmills @mallman Did you try to use a debugger with your M1 Mac for reproducing that case? It would be helpful to determine the area of the code which is triggering this runtime error. Maybe we are using the `std::regex` incorrectly?

I have not looked into the regex failure. I have run `exiv2 -pa test/data/_DSC8437.exv` through the Xcode debugger on arm64 architecture and found something that I don't understand. Granted my understanding of C++ is limited, but as I understand it the ubsan is complaining because it's expecting an object of type `Exiv2::Metadatum` and finding an object of type `Exiv2::Exifdatum` instead. This puzzles me because I think `Exiv2::Exifdatum` is a subtype of `Exiv2::Metadatum`. In my experience from other programming languages with type inheritance, an object is assignable to any variable of its supertypes. I may be missing the need for some kind of explicit cast here.

The ubsan aborts here:

https://github.com/Exiv2/exiv2/blob/b626f241c10de77dd20765cd5f396a7fe8b39b44/app/actions.cpp#L488

which is called from within this loop:

https://github.com/Exiv2/exiv2/blob/b626f241c10de77dd20765cd5f396a7fe8b39b44/app/actions.cpp#L418-L424

**Comment by clanmills** [comment]
I've remembered why I didn't run this in the debugger.  Xcode is unable to build exiv2 with `-DEXIV2_TEAM_USE_SANITIZERS=1`.  Fails to link.  Can't find the ASAN libraries (or something like that).

I've never learnt gdb.  Maybe I can use Xcode or Eclipse to step a debug build created in the terminal.   Maybe we can add something to CMake to compile app/actions.cpp without the ASAN settings and evade this issue.  Similar magic for src/version.cpp to avoid the issue in #2115.

**Comment by piponazo** [comment]
Taking a look at the code I do not see anything wrong with it, it looks like perfectly valid usage of inheritance in C++. When I have some time I can try to also compile and run this on a Rasperry Pi which also has ARM processor, just to check if the problem is also reproducible there.

**Comment by clanmills** [comment]
This is a corner case.  Its a "engineering build" _(not a production build)_ on the arm64/m1 processor.  It's the corner in which Michael and I work.  The other 99% of users will never see this.

I can't remember why we became interested in the ASAN build this week.  We were dealing with Nikon/MakerNote/AF and wandered off the trail into this jungle.  @mallman Can you remember why we are using ASAN at the moment.

**Comment by kevinbackhouse** [comment]
This looks like a bug in the clang UBSAN implementation on Mac. I can't reproduce it on Linux.

**Comment by mallman** [comment]
I don't know C++ well enough to have a good opinion on whether this is some kind of obscure problem in exiv2's use of C++ or a bug (false positive) in the arm64 version of clang's ubsan. If it's the latter, I wonder if we should really even track this issue in this project?

> Maybe I can use Xcode or Eclipse to step a debug build created in the terminal.

That's what I've done (using the Xcode-based debugger).

> Can you remember why we are using ASAN at the moment.

As I recall, you saw a bunch of test errors when you turned it on and ran the unit tests for #2110.

**Comment by clanmills** [comment]
I've remembered why we got interested in ASAN.  I was speculated that if I "doctored" `test/data/_DSC8437.exv` by changing the count in tag 0x7b from 84 to 6, I could cause an ASAN error because the code would access bytes beyond 6.  In fact, the code is safe, however I develop on a MacBookPro/m1.

We've wandered off track here.  The ASAN errors are coming from the ASAN/m1 build.  @kevinbackhouse has offered the opinion that the ASAN/m1 build issue is caused by the tools and not by the exiv2 code.  I agree with Kev. 

I opened both this issue and #2115.  Neither have been really "solved" at this time, however the issue is documented in  #2117 which I will leave open at this time.
